### PR TITLE
FFI: bluechi-tester: initial code

### PR DESCRIPTION
**Pull request by dougsland**
bluechi-tester is a Free From Interfence tool for making sure bluechi-controller has no interference from agents when abusing dbus communication.

Signed-off-by: Michael Engel mengel@redhat.com
Signed-off-by: Douglas Schilling Landgraf dougsland@redhat.com

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

**Comment by dougsland** [comment]
Unit test error not related: 0s
```
Error: Value cannot be null. (Parameter 'ContainerId')
```

**Comment by dougsland** [comment]
Integration Error not related:
```
        + '[' integration-test-local = integration-test-snapshot ']'
        + podman build -f ./containers/integration-test-local -t bluechi-image .
        Trying to pull quay.io/bluechi/integration-test-base:latest...
        Error: error creating build container: initializing source docker://quay.io/bluechi/integration-test-base:latest: reading manifest latest in quay.io/bluechi/integration-test-base: received unexpected HTTP status: 500 Internal Server Error

```

**Comment by engelmi** [comment]
@dougsland Seems like GH actions knows it's Friday. Rerunning unit tests solved it, the integration tests fail due to some issues with quay.io - lets wait a bit and rerun them again.

File: tests/tools/FFI/bluechi-tester/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  bluechi-tester [flags]
+
+Flags:
+  -h, --help                help for bluechi-tester
+  -n, --nodename string     Specify a node name
+  -N, --numbersignals int   Number of signals to be sent (default 1)
+  -s, --signal string       Specify a signal name
+  -u, --url string          Specify a URL
+
+
```
Comment (engelmi) [inline tests/tools/FFI/bluechi-tester/README.md]:
```suggestion
```
These instructions are still from QM. I think you can remove these and, if you want, add a slim `Makefile` for with targets for building (and maybe running).

File: tests/tools/FFI/bluechi-tester/bluechi-tester.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	BcDusInterface = "org.eclipse.bluechi"
+	BcObjectPath   = "/org/eclipse/bluechi/internal"
+	MethodRegister = "org.eclipse.bluechi.internal.Manager.Register"
+)
+
+// object path and signals of internal agents interface
+// see https://github.com/containers/bluechi/blob/main/data/org.eclipse.bluechi.internal.Agent.xml#L81-L116
+const (
+	InternalAgentObject = "/org/eclipse/bluechi/internal/agent"
+	InternalAgentSignal = "org.eclipse.bluechi.internal.Agent."
```
Comment (engelmi) [inline tests/tools/FFI/bluechi-tester/bluechi-tester.go]:
```suggestion
	InternalAgentSignalPrefix = "org.eclipse.bluechi.internal.Agent."
```
What do you think? @dougsland

File: tests/tools/FFI/bluechi-tester/bluechi-tester.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		if signalName == "JobDone" {
+			fmt.Println("Sending " + fmt.Sprint(numberSignals) + " " + signalName + " signal to " + urlValue)
+			// emit a JobDone signal, results in an error log on the controller side since JobID 1 or (N) is unknown
+			for i := 0; i < numberSignals; i++ {
+				err = conn.Emit(InternalAgentObject, InternalAgentSignal + signalName, 1, "fake-state")
+				if err != nil {
+					fmt.Fprintln(os.Stderr, "Failed to emit signal")
+				}
+			}
+		}
```
Comment (engelmi) [inline tests/tools/FFI/bluechi-tester/bluechi-tester.go:105]: Maybe adding an `else` with printing an error, e.g. `Unknown signal name: xyz`, would be good.

**Review by engelmi — COMMENTED** [review]
In general, looks good to me. 
Question that remains is how to provide this application to other projects like QM. That can be tackled in the next PR, though.

File: tests/tools/FFI/bluechi-tester/bluechi-tester.go (hunk 3 above)
Comment (dougsland) [reply tests/tools/FFI/bluechi-tester/bluechi-tester.go:105]: Hi @engelmi, should we remove this code and add the else:
```
 if signalName != "" && signalName != "JobDone" {
                        fmt.Printf("Invalid signal name: %s. Only 'JobDone' is accepted.\n", signalName)
                        os.Exit(1)
                }
```

**Comment by dougsland** [comment]
alright, discussed in the bluechi meeting and will be required to re-implement in python. Closing for now, the good thing is that BlueChi Controller passed in the tests and we have this code as example anyway. See you soon!

**Pull request closed without merging**